src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+        self.transformer = nn.ModuleDict(
+            {
+                "wte": nn.Embedding(config.input_vocab_size, config.hidden_size),
+                "wpe": nn.Embedding(config.block_size, config.hidden_size),
+                "drop": nn.Dropout(config.dropout),
+                "h": nn.ModuleList([BarkBlock(config, idx, is_causal=True) for idx in range(config.num_layers)]),
+                "ln_f": LayerNorm(config.hidden_size, bias=config.bias),
+            }
+        )
+        self.lm_head = nn.Linear(config.hidden_size, config.output_vocab_size, bias=False)
+
+    def get_input_embeddings(self):
+        return self.transformer.wte
+
+    def set_input_embeddings(self, new_embeddings):
+        self.transformer.wte = new_embeddings
+
+    def get_num_params(self, non_embedding=True):
+        """
+        Return the number of parameters in the model. For non-embedding count (default), the position embeddings get
+        subtracted. The token embeddings would too, except due to the parameter sharing these params are actually used
+        as weights in the final layer, so we include them.
+        """
+        n_params = sum(p.numel() for p in self.parameters())
+        if non_embedding:
+            n_params -= self.transformer.wte.weight.numel()
+            n_params -= self.transformer.wpe.weight.numel()
+        return n_params
+
+    def prepare_inputs_for_generation(self, input_ids, past_key_values=None, **kwargs):
+        input_embeds = kwargs.get("input_embeds", None)
+
+        # only last token for inputs_ids if past is defined in kwargs
+        if past_key_values:
+            seq_len = input_ids.shape[1]
+            input_ids = input_ids[:, [-1]]
+
+            if input_embeds is not None:
+                # input_embeds have already been used and is not required anymore
+                input_embeds = None
+
+        attention_mask = kwargs.get("attention_mask", None)
+        position_ids = kwargs.get("position_ids", None)
+
+        # ensure that attention_mask and position_ids shapes are aligned with the weird Bark hack of reducing sequence length on the first forward pass
+        if past_key_values is None:
+            if input_embeds is not None:
+                seq_len = input_embeds.shape[1]
+            else:
+                seq_len = input_ids.shape[1]
+
+        if attention_mask is not None:
+            attention_mask = attention_mask[:, :seq_len]
+        if position_ids is not None:
+            position_ids = position_ids[:, :seq_len]
+
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past_key_values:
+                position_ids = position_ids[:, -1].unsqueeze(-1)
+        else:
+            position_ids = None
+
+        if input_embeds is not None:
```

Comment:
Actually, it is not ok to always return `input_embeds`. 
`input_embeds` has to use in priority w.r.t `input_ids` during the first forward pass of the `.generation`. In any other cases, it should not be the case because it's in `input_ids` that the generated tokens are added. Really, it's about reconciliating Bark weird trick of summing embeddings before the first forward pass, with the `transformers` library

Reply:
Alright I see, thanks for the explanation! So this handles the case where:
* `input_embeds` is not `None`
* `past_key_values` is not `None` (i.e. use cache) 

-> then we know to set  `input_embeds=None` since we've generated new ids (as per our `past_key_values` being not `None`)

But what if:
* `input_embeds` is not `None`
* `past_key_values` is always `None` (i.e. no cache) 

-> then we always pass the `input_embeds` to the model, even if we've generated new `input_ids`?

Reply:
Yes that's right, it's a shortcoming of this code. I'm not sure how to bypass this issue without adding an additional input variable (a boolean). What do you think?

Reply:
Anyways, the standard output is `logits`, so always passing `input_embeds` does not make that much sense, right? 
An user that knows its deal enough to always pass `input_embeds` to the model will surely add the additional `input_id` to `input_embeds` through the embedding layer.

Reply:
But it's true that in the case of passing `input_embeds` and `use_cache = False` to `.generate`, the model won't generate anything. So I have to find a way to bypass this case...

Reply:
I think the way to go is to add a condition that ensures that `input_embeds` is passed to `forward` only when:

- it is `not None`
- `past_key_values is None`
- and `use_cache is True` 
with the last two conditions to make sure `input_embeds` is used only once, and to make sure that it won't be use in the case `use_cache=False`, i.e when no cache.